src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1196,6 +1197,25 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
 
                 tokens.Add(modTok);
             }
+
+            bool isStructOrRecordKeyword(SyntaxToken token)
+            {
+                if (token.Kind == SyntaxKind.StructKeyword)
+                {
+                    return true;
+                }
+
+                if (token.ContextualKind == SyntaxKind.RecordKeyword)
+                {
+                    // This is an unusual use of LangVersion. Normally we only produce errors when the langversion
+                    // does not support a feature, but in this case we are effectively making a language breaking
+                    // change to consider "record" a type declaration in all ambiguous cases. To avoid breaking
+                    // older code that is not using C# 9 we conditionally parse based on langversion
+                    return IsFeatureEnabled(MessageID.IDS_FeatureRecords);
```

Comment:
>IDS_FeatureRecords [](start = 54, length = 18)

Shouldn't we use "record struct" feature? Perhaps we can avoid the breaking change and the language version check if we simply do a look ahead for the `struct` keyword after the `record`?

Reply:
I think this is the correct version to use because it's the version when `record` became a keyword in type declarations and references.

You're correct that there is a behavior change, but I think it's a breaking change that we intended for C# 9 already (`record` is no longer an accepted type name):
```C#
ref record C { get => throw null; set => throw null; } // previously parsed as a top-level ref-returning property, but now parses as a record with errors in members

public class record { } // no longer a proper type in C# 9
```

Note that other type references to `record` where already disallowed. Example: `record x = null; // parses as broken type declaration, not local declaration in C# 9`

---
In reply to: [580447902](https://github.com/dotnet/roslyn/pull/51252#discussion_r580447902) [](ancestors = 580447902)